### ISSUE TITLE
ActiveRecord relation with multiple database connections.

### ISSUE THREAD

**Issue by shranet** [yiisoft/yii2#15012]
I configured multiple database connections in console. Both databases are the same structure but different data.

For example: Table "orders":

In db1:

| id    | order_date 
| ---  | --------------------- 
| 1     | 2017-10-21 00:00:00

In db2:

| id    | order_date 
| ---  | --------------------- 
| 1     | 2018-11-21 01:02:03

Table "order_products":
In db1:

| id | order_id | product_id
| -- | -------- | ----------
| 1   | 1 | 1
| 1   | 1 | 2

In db2:

| id | order_id | product_id
| -- | -------- | ----------
| 1   | 1 | 3

Two models:
1. "Orders" model:
    
    ...
    public functon getOrderProducts()
    {
        return $this->hasMany(OrderProducts::className(), ["order_id" => "id]);
    }

2. OrderProducts model.

Test:

    $order1 = Orders::find()->with("orderProducts")->where(["id" => 1])->one();
    $order2 = Orders::find()->with("orderProducts")->where(["id" => 1])->one(Yii::$app->db2);
    echo $order1->order_date." ".$order2->order_date."\n";
    echo count($order1->orderProducts)." ".count($order2->orderProducts)."\n";
    //output
    //2017-10-21 00:00:00 2018-11-21 01:02:03
    //2 2

On second query relation used default database instance.

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.12
| PHP version      |  5.6.26
| Operating system | MacOS Sierra 10.12.6

**Comment by bscheshirwork** [comment]
@shranet var_dump($order1->orderProducts, $order2->orderProducts) pls

**Comment by bscheshirwork** [comment]
@shranet 
I think in you case `db` of `OrderProducts` gets from class definition

**Comment by shranet** [comment]
Code:

    
    public function actionIndex($message = 'hello world')
    {
        $order1 = Orders::find()->with("orderProducts")->where(["id" => 1])->one();

        echo "Default DB\n\n";
        print_r($order1->attributes);
        $this->printOrderProducts($order1->orderProducts);

        $order2 = Orders::find()->with("orderProducts")->where(["id" => 1])->one(Yii::$app->db2);

        echo "---------\nDb2\n\n";
        print_r($order2->attributes);
        $this->printOrderProducts($order2->orderProducts);

        echo "Db DSN: ".Yii::$app->db->dsn."\n";
        echo "Db2 DSN: ".Yii::$app->db2->dsn."\n";
    }

    private function printOrderProducts($list)
    {
        echo "Order products\n";
        foreach($list as $n => $row) {
            print_r($row->attributes);
        }
    }
    

Result:

    Default DB

    Array
    (
        [id] => 1
        [order_date] => 2017-10-20
    )
    Order products
    Array
    (
        [id] => 1
        [id_order] => 1
        [id_product] => 1
    )
    Array
    (
        [id] => 2
        [id_order] => 1
        [id_product] => 2
    )
    ---------
    Db2

    Array
    (
        [id] => 1
        [order_date] => 2018-11-22
    )
    Order products
    Array
    (
        [id] => 1
        [id_order] => 1
        [id_product] => 1
    )
    Array
    (
        [id] => 2
        [id_order] => 1
        [id_product] => 2
    )
    Db DSN: pgsql:host=127.0.0.1;dbname=test1
    Db2 DSN: pgsql:host=127.0.0.1;dbname=test2

SQL:
    
    Database 1

    CREATE TABLE orders (
        id serial,
        order_date date
    );

    CREATE TABLE order_products (
        id serial,
        id_order integer,
        id_product integer
    );

    INSERT INTO orders (id, order_date) VALUES (1, '2017-10-20')
    INSERT INTO order_products(id, id_order, id_product) VALUES (1, 1, 1);
    INSERT INTO order_products(id, id_order, id_product) VALUES (1, 1, 2);

    Database 2

    CREATE TABLE orders (
        id serial,
        order_date date
    );

    CREATE TABLE order_products (
        id serial,
        id_order integer,
        id_product integer
    );

    INSERT INTO orders (id, order_date) VALUES (1, '2017-11-22')
    INSERT INTO order_products(id, id_order, id_product) VALUES (1, 1, 3);

The problem is, the second query relation returns database1 values!

**Comment by bscheshirwork** [comment]
@shranet what result is `$order2->orderProducts[0]->getDb()` ?

I think ff your use lazy load your $db for relation model gets from relation class definition, like this:
https://github.com/yiisoft/yii2/blob/ce65e24d1988f00e57ceaa7493632ee4f3aca776/framework/db/ActiveRecord.php#L126-L135

**Comment by bscheshirwork** [comment]
@shranet try it

```php
...
    $order1 = Orders::find()->joinWith("orderProducts")->where(["id" => 1])->one();
...
    $order2 = Orders::find()->joinWith("orderProducts")->where(["id" => 1])->one(Yii::$app->db2);
...
```

**Comment by shranet** [comment]
$order2->orderProducts[0]->getDb()->dsn returns "pgsql:host=127.0.0.1;dbname=test1".

    $order1 = Orders::find()->alias("t")->joinWith("orderProducts")->where(["t.id" => 1])->one();
    $order2 = Orders::find()->alias("t")->joinWith("orderProducts")->where(["t.id" => 1])->one(Yii::$app->db2);

returns same result.

If I using this in OrderProduct,

     public static function getDb() 
     { 
         return Yii::$app->getDb(); 
     } 

how I can detect Order ActiveRecord used db instance or db1?

**Comment by bscheshirwork** [comment]
Hm, so smart ActiveRecord... I think we see 2 different db and split query on 2 query. Or we have bug ^^

You can see query log? You can use [mysql-proxy](https://github.com/bscheshirwork/docker-mysql-proxy) or some else?

So... You can try simple hardcore way: 
```php
class OrderProductDb2 extends OrderProduct {
 public static function getDb() 
 { 
     return Yii::$app->db2; 
 } 
}
```
```php
class Orders{
...
 public function getOrderProductsDb2(){
return ...hasMany(OrderProductDb2::class ...
...
}
}
```
```php
$order2 = Orders::find()->alias("t")->joinWith("orderProductsDb2")->where(["t.id" => 1])->one(Yii::$app->db2);

```

**Comment by bscheshirwork** [comment]
@shranet how about
```php
var_dump(Orders::find()->alias("t")->joinWith("orderProductsDb2")->where(["t.id" => 1])->createCommand(Yii::$app->db2)->getRawSql() );
```

**Comment by shranet** [comment]
var_dump(Orders::find()->alias("t")->joinWith("orderProducts")->where(["t.id" => 1])->createCommand(Yii::$app->db2)->getRawSql() );

returns

    'SELECT "t".* FROM "orders" "t" LEFT JOIN "order_products" ON "t"."id" = "order_products"."id_order" WHERE "t"."id"=1' (length=116)

SQL Log:

|Query|Time|
|-------|------|
|SELECT "t".* FROM "orders" "t" LEFT JOIN "order_products" ON "t"."id" = "order_products"."id_order" WHERE "t"."id"=1|13:22:26.400|
|SELECT * FROM "order_products" WHERE "id_order"=1|13:22:26.405|
|SELECT "t".* FROM "orders" "t" LEFT JOIN "order_products" ON "t"."id" = "order_products"."id_order" WHERE "t"."id"=1|13:22:26.416|
|SELECT * FROM "order_products" WHERE "id_order"=1|13:22:26.420|

**Comment by bscheshirwork** [comment]
So... As we can see in this query... Split on 2 independent query.
@samdark 

interesting 
`$query->one($notnull)`
```php
    public function one($db = null)
    {
        if ($this->emulateExecution) {
            return false;
        }

        return $this->createCommand($db)->queryOne();
    }
```
difference logic of 
```php
$query->createCommand($notnull)->getRawSql()
```
?

**Comment by cebe** [comment]
> On second query relation used default database instance.

this is expected, activerecord will always call `static::getDb()` for queries where you did not explicitly specify the database. If you want the relation from the other database, you have to specify it for that query too:

```php
$order2 = Orders::find()->where(["id" => 1])->one(Yii::$app->db2);
$relatedItems = $order2->getOrderProducts()->all(Yii::$app->db2);
```

Its currently not possible to specify the database for queries in `with`. This is mainly because you can have relations defined over different databases, so the database used for the query of the original record is considered irrelevant for the relational query.

**Comment by samdark** [comment]
Won't be implemented in 2.0.

**Comment by Twilio-taskrouter232** [comment]
> So what is the issue?





https://phps.space/questions/exploring-the-data-within-yii2
